File: leetcode/Go/backtracking/79/79.go
package main

func exist(board [][]byte, word string) bool {
	exists := false
	var findWord func(x, y, wordIndex int)
	findWord = func(x, y, wordIndex int) {
		if wordIndex == len(word) {
			exists = true
			return
		}
		// If we are in the limits
		if x < 0 || x >= len(board) || y < 0 || y >= len(board[0]) || board[x][y] != word[wordIndex] {
			return
		}
		temp := board[x][y]
		// Setting a different value temporary
		board[x][y] = '#'
		// Up
		findWord(x, y+1, wordIndex+1)
		// Down
		findWord(x, y-1, wordIndex+1)
		// Left
		findWord(x-1, y, wordIndex+1)
		// Right
		findWord(x+1, y, wordIndex+1)
		board[x][y] = temp
	}
	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			findWord(i, j, 0)
			if exists {
				return true
			}
		}
	}
	return exists
}

File: leetcode/Python/backtracking/79.py
from typing import List


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        exists = False
        directions = [(0,1), (1, 0), (0, -1), (-1, 0)]
        def backtracking(i, j, wordIndex) -> bool:
            if board[i][j] != word[wordIndex]:
                return False
            if wordIndex == len(word):
                return True

            tmpVal = board[i][j]
            board[i][j] = "#"

            for direction in directions:
                x, y = direction
                x, y = x + i, y + j
                if  0 <= x < len(board) and  0 <= y < len(board[0]):
                    if backtracking(x, y, wordIndex+1):
                        return True

            board[i][j]= tmpVal
            return False 
            
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    exists = backtracking(i, j, 0)
                    if exists:
                        return True
                    
        return exists
        
